### DISCUSSION TITLE
Is _generate_cache_key intended to be safe for public use?

### DISCUSSION THREAD

**Discussion by BrandonWiebe** [sqlalchemy/sqlalchemy#10456] [Usage Questions] [answered]
Hi there,

I'd like to make use of sqlalchemy's statement compilation cache key to memoize a function that takes in statement objects as input, but as far as I can tell the only way to access said cache key is to call the `_generate_cache_key` method on a given statement. Given the leading underscore in the method name, I'm not sure if I should really be calling this method from these other contexts, however. 

The documentation mentions this method in a few different places, but generally it seems to only reference it when illustrating the internals of the statement compilation system. However, I can see that it is used in the [dogpile cache example](https://docs.sqlalchemy.org/en/20/_modules/examples/dogpile_caching/caching_query.html) that's provided in the docs to compose a key for a SQL results cache. So, I was wondering if it's safe to use this internal method in the context of function memoization, or is this approach not recommended?

**Comment by zzzeek** [comment, marked as answer]
hi - 

I think it's in the dogpile.cache example because it is very useful in that context, and it's obviously being used in an extremely carefully vetted context.   the caching thing, as you might imagine, was ridiculously hard to get right, and it took a few years to iron out all the issues.  so making this attribute completely public out of the gate would have been dangerous as we didnt know how the caching was going to play out and what changes might have been needed.

I'm still not in a hurry to make this attribute "fully public", because that implies things like this key should be generally available from all kinds of constructs (which it isn't right now, for example DDL and some dialect-specific constructs dont have it) and of course the format of the key can change, though this it not expected in the near term.     right now the cache key consists of a gigantic tuple that is a traversal of the whole object, which then via Python's ability for tuple to produce a `hash()` value, works at that level.  some future, likely cython-based system that can just flatten down the traversal into a hash number directly might be desirable.

So for now I'd characterize this API as "stable, dont rely on the structure of the key portion, use judiciously as this is at best semi-public", and we aren't right now supporting changes or enhancements to this key to support other use cases.

**Comment by BrandonWiebe** [comment]
Excellent, thanks for the thorough answer!
